src/dragonball/src/tracer.rs

```diff
@@ -0,0 +1,176 @@
+    /// Subscriber guard. The lifetime is as same as tracer.
+    sub_guard: Option<DefaultGuard>,
+    /// Tracing enabled flag.
+    enabled: bool,
+    /// Dragonball tracing root span.
+    root_span: Option<EnteredSpan>,
+}
+
+/// The fields should always be private and the methods should be exposed in a safe.
+unsafe impl Send for DragonballTracer {}
+
+impl Default for DragonballTracer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DragonballTracer {
+    /// Create a dragonball tracer instance.
+    pub fn new() -> Self {
+        Self {
+            subscriber: Arc::new(NoSubscriber::default()),
+            sub_guard: None,
+            enabled: false,
+            root_span: None,
+        }
+    }
+
+    /// Return whether the tracing is enabled, enabled by [`setup_tracing`]
+    pub fn enabled(&self) -> bool {
+        self.enabled
+    }
+
+    /// setup tracing feature.
+    pub fn setup_tracing(&mut self, config: TraceInfo) -> Result<(), TraceError> {
+        if self.enabled {
+            warn!("Tracing is already enabled");
+            return Err(TraceError::AlreadyEnabled);
+        }
+        self.enabled = true;
+        self.subscriber = config.subscriber;
+
+        // set the default subscriber for the current thread
+        self.sub_guard = Some(tracing::subscriber::set_default(self.subscriber.clone()));
+        self.root_span = Some(trace_span!("dragonball-root").entered());
+        info!("Enable tracing successfully");
+        Ok(())
+    }
+
+    /// end tracing for application.
+    pub fn end_tracing(&mut self) -> Result<(), TraceError> {
+        if self.enabled {
+            // root span must drop before the default subscriber guard is dropped.
+            if let Some(span) = self.root_span.take() {
+                span.exit();
+            }
+            if let Some(guard) = self.sub_guard.take() {
+                drop(guard);
+            }
+            self.enabled = false;
+            info!("Ending tracing successfully");
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+    use std::sync::Mutex;
+
+    use opentelemetry::global;
+    use tracing::subscriber::set_global_default;
+    use tracing::{instrument, trace_span, Subscriber};
+    use tracing_opentelemetry::OpenTelemetryLayer;
+    use tracing_subscriber::prelude::*;
+    use tracing_subscriber::Registry;
+
+    use crate::tracer::{DragonballTracer, TraceInfo};
+
+    const DEFAULT_JAEGER_URL: &str = "http://localhost:14268/api/traces";
+
+    struct TestLayer {
+        pub tracer: Arc<Mutex<DragonballTracer>>,
+    }
+    impl TestLayer {
+        pub fn new() -> Self {
+            Self {
+                tracer: Arc::new(Mutex::new(DragonballTracer::new())),
+            }
+        }
+    }
+
+    #[instrument]
+    fn tracing_fn_test() {
+        trace_span!("span in fn");
+    }
+
+    #[test]
+    fn test_tracing_in_lib() {
```

Comment:
Is that more like a `integration test` instead of `unit test`?
Will that be too much to create a JAEGER service every time we run unit test?

Reply:
Yes, this is more like an example and I have removed it. Actually, `opentelemetry_jaeger`  just creates a request to export the tracing info to JAEGER.